### PR TITLE
Error for groupshared outside of compute

### LINKED ISSUE #2603: Frontend needs to disallow RayQuery from being placed in groupshared memory or wave ops (e.g. WaveReadLaneAt()).
RayQuery in groupshared is too big a can of worms to try to get working (at least for now).   Could revisit in the future.

For now, compiler needs to just disallow it.

### PR DESCRIPTION
**Pull request by pow2clk**
groupshared variables shouldn't be allowed outside of compute and
compute-like shaders. This adds a Sema and validation error when such
variables are used.

Add shader mask where groupshared are used

Correct several tests that used groupshared incorrectly

Fixes #2603, #2677

### CODE REVIEW

File: tools/clang/lib/Sema/SemaHLSL.cpp
Comment (pow2clk) [inline tools/clang/lib/Sema/SemaHLSL.cpp:12229]: I have already convinced myself that emitting an error here isn't worth it, but I want a second opinion.

The problem with emitting the error here is that any shader that has any groupshared variables at all will produce an error if it is ever built for a target that doesn't support it. That's something a number of our tests (many written by yours truly) did and it's likely that there are other such shaders out in the wild. Since we produce the error in validation only when the groupshared variable is actually used by a disallowed shader stages, it's a more accurate error that allows us to irritate people as little as possible.

**Comment by pow2clk** [comment]
This change may look intimidating, but the vast majority of it is simple fixes to tests

**Comment by pow2clk** [comment]
The failing SPIR-V tests are pixel shaders that make use of groupshared. The new failures are expected.

I'm having some trouble figuring out how some of them passed in the first place. In particular, spirv.debug.opline.function.hlsl has loads of OpLine checks, but even with ToT only produces 3 or so. Yet it somehow passes. I wonder if the debug information which includes the source file comments is matching those somehow. The regexp usually bypasses that though so I'm unsure.

In addition, some of the tests make use of ops only available to pixel shaders along with groupshared memory. For example, spirv.debug.opline.intrinsic.hlsl uses clip() and ddx() and also groupshared memory. Incidentally, trying to call clip() within a compute shader produces an error that encourages the reader to file a github issue. using ddx() doesn't produce any error or warning at all even though that functionality was only just added with SM 6.6. For this case, the solution is either to split the shader, lose the groupshared, or lose the clip. Since I didn't write these tests, I don't know what's preferred.

Given my various uncertainties, I call out to @ehsannas or @jaebaek to give me some guidance please!

**Comment by ehsannas** [comment]
Thanks for the heads up @pow2clk. @jaebaek will push a commit to this PR to fix these. The debug info tests are supposed to focus their checks on filename/line/column/scope/etc information, and so the shader stage and use of groupshared in them is immaterial (sad that they weren't written more thoughtfully in the first place though).

@jaebaek can you also please double check the different issues that Greg described above and make sure we fix them? (unblock this PR from landing first, and make separate PRs for other fixes, if any). Thanks!

**Comment by jaebaek** [comment]
Thank you for heads up. I just prepared a PR to fix SPIR-V part failures: https://github.com/microsoft/DirectXShaderCompiler/pull/3474.

I will open a bug about the error message of `clip()` and `ddx()`.

File: tools/clang/lib/Sema/SemaHLSL.cpp
Comment (tex3d) [reply tools/clang/lib/Sema/SemaHLSL.cpp:12229]: No we can't have an error here, because you are basing this on the top-level target and applying this to all code.  It's common to compile different shaders out of the same source, so this would break perfectly legal scenarios.  If we want an error before validation, it needs to be late, when we know it's actually used in the shader, after DCE.

**Review by tex3d — CHANGES_REQUESTED** [review]
I have one comment where we need to remove the code that blocks using any groupshared decls at all in the HLSL source from which you compile a non-compute shader.  This is wrong.  It's fine as long as the shader doesn't actually use it, since it's common to compile multiple shaders from the same source.

**Review by tex3d — APPROVED** [review]
LGTM!

**Pull request merged**
